library/src/engine/engine.ts

```diff
@@ -25,23 +25,25 @@ const plugins: AttributePlugin[] = []
 const actions: ActionPlugins = {}
 const watchers: WatcherPlugin[] = []
 
+// Map of cleanup functions by element ID, keyed by a dataset key-value hash
+const removals = new Map<string, Map<number, OnRemovalFn>>()
```

Comment:
@delaneyj note that I changed the key from an element to an element ID to allow for rehashing in the `mergeFramgents` plugin.